_posts/2021-01-25-troubleshooting-puppetized-dsc-modules.md

```diff
@@ -0,0 +1,364 @@
+---
+layout: post
+title: "Troubleshooting Puppetized DSC Modules"
+author: michaeltlombardi
+---
+
+Occasionally, something will go wrong when trying to apply a manifest containing Puppetized DSC resources.
+There's a few different ways this can happen:
+
+- There can be a misgenerated type, in which case the Puppet representation of the DSC resource does not match the resource's API surface.
+  This can be mismatched enums or types—in these cases, the issue is with this builder and a bug will need to be submitted for this project so the Puppetized module can be rebuilt with the patched builder.
+- There can be an error in the Ruby code that translates the DSC resource back and forth between Puppet and PowerShell, in which case the issue is with the base provider that lives in the pwshlib module; Once a bug is submitted for that project and a fix is released, the manifest and Puppetized module should work without any updates except in the dependency pwshlib module.
+- There can be an error in the underlying DSC resource's implementation in PowerShell—in this case, you'll need to submit a bug with the upstream maintainers for that DSC resource's PowerShell module.
+  Once a new release of that module is out a new build of the Puppetized module can happen, vendoring in those updates.
+
+This blog post dives into a little more detail on how to specifically debug and troubleshoot Puppetized DSC Modules to discover which of these error categories you're experiencing and help you file detailed and useful bug reports that should see these bugs resolved more quickly.
+
+## Basics
+
+The first thing to note is that there's actually a [basic troubleshooting guide for `Puppet.Dsc`][troubleshooting-doc].
+That guide already covers finding [type errors][type-errors], [incorrect enums][incorrect-enums], and [invocation errors][invocation-errors] during a Puppet run, so we won't cover those again here.
+
+It also covers running Puppet in **debug** mode, which we need to talk about for the next couple sections on how to validate the conversion of your Puppet manifest to the hash being passed to `Invoke-DscResource` and how you can test that invocation outside of a Puppet run.
```

Comment:
Will the audience know what "conversion of puppet manifest to powershell hash" means? Or should you set this up with a brief introduction of how the translation layer works. Something like "at it's heart, this tool just translates Puppet code to a hash of options passed to `Invoke-DscResource`, so we'll also spend some time showing you how to do that on your own so you can validate it outside of a Puppet run...."